### PR TITLE
Fix nll_backward for negative weights

### PR DESCRIPTION
**Pull request by lezcano**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64572

Fixes https://github.com/pytorch/pytorch/issues/64256
It also fixes an inconsistent treatment of the case `reduction = "mean"`
when the whole target is equal to `ignore_index`. It now returns `NaN`
in this case, consistently with what it returns when computing the mean
over an empty tensor.

We add tests for all these cases.

Differential Revision: [D31116297](https://our.internmc.facebook.com/intern/diff/D31116297)

cc @ezyang @gchanan

### CODE REVIEW

**Comment by Lezcano** [comment]
This has an UBsan problem that I'll fix an explicit instantiation of a `quiet_NaN` and also an XLA test is failing, but it's basically ready for a review @jbschlosser

File: aten/src/ATen/native/LossNLL.cpp
Comment (jbschlosser) [inline aten/src/ATen/native/LossNLL.cpp:186]: Why was this check removed?

**Review by jbschlosser — COMMENTED** [review]
Thanks for the fix! :)

Since the behavior here wrt `reduction='mean'` is different now, it's possible that some edge cases in the wild might break.

Example: a batch full of targets that have `weight=0`. Before, this would return 0, but now it will return NaN. According to the math in the docs, NaN is correct, but that math is contested in #61309, and it's possible returning NaN instead of 0 would be surprising for similar reasons that the way we calculate the mean is surprising to some.

Further, if we change the behavior, we'd need to make corresponding changes in XLA and deal with the UBSAN failure, as evidenced by the failing checks. These aren't hard to do; just requires some coordination with the XLA team and creating a NaN explicitly instead of doing `0/0`.

However, rather than go through with all of this, I think I'd prefer to keep this PR to the minimal fix necessary to correctly handle negative weights for now. We can possibly address the BC-breaking parts in a future PR to keep things contained. Wdyt about this?

File: aten/src/ATen/native/LossNLL.cpp
Comment (Lezcano) [reply aten/src/ATen/native/LossNLL.cpp:186]: As these are structured kernels, all the checks happen there. Pretty sure they forgot to delete this when porting these functions to structured kernels.

**Comment by Lezcano** [comment]
About the BC-breaking. I don't think this is BC-breaking as much as a bug fix. As you mentioned, this is the behaviour in the docs. Even more, at the moment the behaviour is inconsistent across sizes / devices, so that is not something desirable. I say we fix this.

**Comment by jbschlosser** [comment]
@Lezcano Sounds good, I'm convinced, especially in light of the fact that the current behavior is inconsistent across sizes / devices. Let's move forward with the fix.

@JackCaoG FYI it looks like we need similar changes in XLA to get the tests passing.

**Comment by JackCaoG** [comment]
Hi @jbschlosser , sounds good. @Lezcano Feel free to make the change on the pytroch side, xla CI in this pr should fail and youcan open an issue on the pt/xla side. We will follow up.

**Comment by jbschlosser** [comment]
Otherwise, just need the UBSAN fix and we're good to go :)

**Comment by Lezcano** [comment]
@jbschlosser I fixed the UBSAN failures. Now the only CI failures remaining are those from XLA (expected) and a couple that complain about the base of this branch. This should be ready.

**Comment by JackCaoG** [comment]
@Lezcano pt/xla CI failure seems to be related to merge conflict. any chance you can rebase and surface the true error? If you could let me know which test you expect to fail I might be able to repo the failure myself too.

**Comment by Lezcano** [comment]
@JackCaoG rebased.

**Comment by JackCaoG** [comment]
@Lezcano pt/xla test passed, is the new behavior not being tested by the existing test?

**Comment by Lezcano** [comment]
@JackCaoG This PR adds new tests, and note that they are failing in an XLA run now.
@jbschlosser all the other runs are passing, this should be ready to go.

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks :)

**Comment by JackCaoG** [comment]
@Lezcano Thanks, I will take care of it.

**Comment by jbschlosser** [comment]
@jbschlosser has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31116297).

**Comment by JackCaoG** [comment]
@jbschlosser I will update here when the xla change is ready

**Comment by jbschlosser** [comment]
> About the BC-breaking. I don't think this is BC-breaking as much as a bug fix. As you mentioned, this is the behaviour in the docs. Even more, at the moment the behaviour is inconsistent across sizes / devices, so that is not something desirable. I say we fix this.

Hm, I'm now running into internal tests that explicitly depend on the old behavior of returning 0.0 for the "all ignored" case, which makes me hesitate again at changing this behavior. This is for an internal loss function based on NLLLoss that explicitly claims to support missing labels - returning NaNs here doesn't seem nice for this case.

If PyTorch had a nice way to indicate "no gradients" when computing the mean of an empty set, we wouldn't have to choose between NaN (breaks training) and 0.0 (allows training to continue, but not mathematically correct and may have undesirable side effects).

cc @cpuhrsch: FYI `NLLLoss` / `CrossEntropyLoss` are a couple more cases where we allow ad-hoc "masking" via `ignore_index`.

I do agree it's undesirable to have inconsistency across sizes / devices - do you mind pointing out the current discrepancies and maybe we could at least fix those?

File: test/test_nn.py
Comment (albanD) [inline test/test_nn.py:17340]: Changing this is BC-breaking no? 

Is it really worth it?

**Comment by Lezcano** [comment]
The state of things is the following:

`nll_loss` has about:
- 3x kernels to cover for each of the reductions
- 2x kernels dealing for inputs of size 1 or 2 and for inputs of size 4
- 2x kernels CPU / CUDA
- 2x kernels forward / backward

= About 24 kernels roughly. 

Unsurprisingly, there are quite a few divergences between these 24 kernels. This PR addresses the following problems:

In the 4 that are forwards with `reduction="mean"`, you may have an input with zero elements just because the input has zero elements or because the all the elements in the target were equal to `ignore_index`. In the first case, the function returns zero for some inputs and NaNs for other: It returns zeros for cpu tensors and NaNs for CUDA tensors. This PR fixes this behaviour to return always NaN, as it's the documented behaviour. Those tests were wrong, and any code that relies on that behaviour is flaky, at the very least, and at worst it has bugs.

When the `ignore_index` was equal to all the indices in , all returned zero, but of course, this is wrong in view of the previous behaviour. This PR fixes this as well.

**Comment by albanD** [comment]
While I agree that the "mean" operation semantic would be to return "NaN". But here we are implementing "nll" and in this case, I actually think that "0" is a good identity element.
That would also match what users expects when this happens during their training.

My vote would be to make it consistently return "0".

**Comment by jbschlosser** [comment]
Thanks for the the additional context! I definitely agree that the state of things is not good, and the inconsistencies should ideally be cleaned up.

However, wrt the internal code, I'm worried about one case in particular: the "all ignored" case with a mean reduction on CPU.
* Training is done on CPU (for better or worse)
* The internal loss function explicitly ignores things with missing labels as a feature, making "all ignored" the case to happen in practice rather than "empty input"
* Mean reduction is used in practice and is the one with ambiguities regarding empty set handling

The fact of the matter is that the behavior for the case above will be changed if we fix the inconsistencies as currently proposed in the PR, and I'm wary of doing that since it's explicitly depended upon. Of course, following from the inconsistencies you pointed out, leaving things as-is will cause problems if training is ever moved to CUDA.

It's possible to address inconsistencies in a non-breaking way by consistently producing 0 instead of NaN for the forward pass with 0 gradients during the backward. I argue that an even better solution would be to have proper masking semantics to explicitly handle fully-masked-out inputs, producing "no gradients" for the backward pass and allowing training to continue for these case without divergence due to NaNs (but we don't have this in PyTorch yet).

**Comment by ngimel** [comment]
Looks like returning "0" in this case requires either a global sync in the kernel, or launching an additional kernel. Neither is great if the purpose is just to fix some edge gradgradcheck case that no one ever encountered. 
Also, the correct value to return is `nan`, and saying that `mean` of zero elements becomes 0 just because it's part of nll loss is not very consistent.

**Comment by Lezcano** [comment]
I pushed a version that returns zeros when all the inputs are ignored. Now, there is quite a bit of specific code to do this, so I do agree that it'd be better and more consistent to return `NaN`s. I'll edit the code later on provided that those internal tests can be moved to accommodate this behaviour.

For reference, `nll_loss` (and other loss functions) could do with a good rewrite in terms of TI for maintenance. At the moment, `nll_loss` encompasses more than 2000 lines of code, where the code tends to be repeated between 12 and 15 times. Making anything consistent is a real pain.

This function even has several implementations of exactly the same logic on GPU. See for example
https://github.com/pytorch/pytorch/blob/e155e7520fc65168e7fdf3706aacb12bb1e087bb/aten/src/ATen/native/cuda/Loss.cu#L226-L272
and
https://github.com/pytorch/pytorch/blob/e155e7520fc65168e7fdf3706aacb12bb1e087bb/aten/src/ATen/native/cuda/NLLLoss2d.cu#L67-L111

A great way of tackling this problem and making all the losses in PyTorch consistent, fast and maintainable would be to have a general `TensorIterator` + code for the reductions similar to what we do in AdvancedIndexing. This way, each particular loss and their derivative would simply be implemented via a couple lambdas and some tests.

**Comment by ngimel** [comment]
Thanks @lezcano, let's try to see if we can go with a more consistent approach of returning nans. 
As for the rewrite, sure, that's a good idea, but probably there are more hi pri things to do (although I fully agree re: maintenance cost). 
Keep in mind that you don't want to have kernels handling general reductions, that's typically too hard, so nn kernels are usually limited to supporting only contiguous tensors/tensors with fixed stride at reduction dimension, and not arbitrarily strided. Arbitrary strides can be handled by copies.

**Comment by Lezcano** [comment]
In this case the reductions act on the whole tensor and we just have 3 possible reductions
- `none` this one's easy
- `sum`: equivalent to `loss(x,y, "none").sum()` 
- `mean`: this one is a bit inconsistent, but in general it's like `loss(x,y, "none").mean()`

I believe that doing generic kernels with not too much boilerplate should be feasible, but yes, it would take quite some time, and the deprecations to make all of these consistent would be painful (as any deprecation on functions that are used widely really).

**Comment by albanD** [comment]
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31116297).

**Comment by albanD** [comment]
The xla test failure seems real. Do we need a sister PR on the xla repo?

**Comment by JackCaoG** [comment]
@albanD I will work on a fix. Already have a prototype, just need to fix some additional failure on the xla end.

**Comment by albanD** [comment]
I'm not very familiar with the joint landing process.
Does that mean that this one can be landed now (or soon when internal tests are done). Or should I wait for your green light?

**Comment by JackCaoG** [comment]
@albanD Please give me 1~2 days, I will update here when xla pr is ready. After that I will merge the xla pr when pytorch pr is merged.

**Comment by JackCaoG** [comment]
@albanD @Lezcano For nll_loss, should grad remain 0 if the forwarding pass return `nan`. Here is an example
```
m = nn.LogSoftmax(dim=1)
loss = nn.NLLLoss(ignore_index=1)

input = torch.randn(5, 2, requires_grad=True, device=device)
target = torch.ones(5, device=device).to(torch.int64)
output = loss(m(input), target)
print(output)
output.backward()
print(input.grad)
```

after my change, CPU returns
```
m = nn.LogSoftmax(dim=1)
loss = nn.NLLLoss(ignore_index=1)

# input is of size N x C = 3 x 5
input = torch.randn(5, 2, requires_grad=True, device=device)
# each element in target has to have 0 <= value < C
target = torch.ones(5, device=device).to(torch.int64)
output = loss(m(input), target)
print(output)
output.backward()
print(input.grad)
```

and xla returns
```
tensor(nan, device='xla:0', grad_fn=<NllLossBackward0>)
tensor([[nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan],
        [nan, nan]], device='xla:0')
```
I can update the xla behavior but want to make sure this is intended.

**Comment by Lezcano** [comment]
I would say that a grad of zero is fine, as it is the fastest one to return. The function is not differentiable at these points (it's not even well-defined) so we are free to return anything we want. As such, we can choose to return whatever's fastest for the computations.

In this case, we can return zeros if all the elements in target were ignored, and inf or nan if the sum of the weights was zero (inf if the `grad_output` was not zero. I'll push an update making sure that this is what we do :)

**Comment by Lezcano** [comment]
For what is worth, I have updated the code for the derivatives to make it more homogeneous. It should now be obvious that all the paths in both CPU and CUDA return what was described here.

**Comment by albanD** [comment]
Thanks!

**Comment by albanD** [comment]
@JackCaoG ok all the internal tests are good. So ready on my end. Let me know when it's good for you!

**Comment by JackCaoG** [comment]
@albanD xla pr is ready at https://github.com/pytorch/xla/pull/3144#pullrequestreview-769344716. I will merge it after this pr being merged.

**Comment by albanD** [comment]
Thanks!
Just clicked land.

**Pull request closed without merging**
